### ISSUE TITLE
Add support for "cubical type theory" under flag --cubical

### ISSUE THREAD

**Issue by Saizan** [agda/agda#1703]
Considering there has been an actual type system with conversion rules for a while:
  http://www.cse.chalmers.se/~coquand/face.pdf

It would be interesting to have a proper proof assistant which implements that theory, to gain support for functional extensionality, bisimilarity-as-equality, quotients, univalence.
##### Partial Elements

The main addition required would be "partial elements", or "systems" as previously called, since the interesting primitives rely on those. This implies changes to both the source and internal syntax.
##### Overloading Lambdas

We might want to reuse lambdas and whitespace for path abstraction and application, instead of `<i> t` and `p @ i`, unless that complicates type inference too much.
##### Pattern Matching and Inductive Families

Concerning pattern matching the flag --cubical should imply --without-K.
That is still not enough to guarantee definitional canonicity at e.g. Bool though, for which we could initially disable inductive families. (The end-goal however would be to have inductive families as parametrized types with embedded equality proofs behind the scenes, while still being able to use dependent pattern matching as we can now. It might need some amount of research and user interface design.)
##### Higher Inductive Types

For higher inductive types we could initially only provide support for quotients and truncations as primitives, considering the theory is not completely settled on them. In "cubicaltt" this means that pattern matching on one of those types also needs to pattern match on the path constructors: we should be able to handle this by extending the grammar for copatterns. 
##### Impact

My impression is that implementing this minimal support for cubical will not have too big of an impact on the codebase, since in the end the Path type is a mix of a dependent function and a singleton type and our conversion checking is already type directed. We might need more type information during reduction though.

I will come up with a more detailed plan in the future, but I opened this issue so that we can collect comments.

**Comment by andreasabel** [comment]
In any case, we should not use the ugly shortcut `t(i 0)` for `t(i = 0)`.

**Comment by Saizan** [comment]
Well, `t(i = 0)` shouldn't really appear in surface syntax, but agreed.

I was looking at .Reduce/.Substitute and it seems we might need to annotate the Var constructor with the type of the variable so that we can reduce e.g. `p @ 0` to `x` in case `p : Path A x y`. I suppose the ReduceEnv actually contains the types of the free vars, but e.g. applyE is a "pure" function.

**Comment by Saizan** [comment]
For anyone following, the development is ongoing at

https://github.com/agda/agda/tree/cubical

with some demonstration/testing code available at

https://github.com/Saizan/cubical-demo

**Comment by xekoukou** [comment]
I would like to build a network library on agda or a similar language. It seems that having bisimilarity as equality would be something that I would like to use.

Could you inform me whether your work will be merged into agda and when? 
I also saw but not read a paper "Guarded Cubical Type Theory: Path Equality for Guarded Recursion"
and was wondering whether it will be implemented here.

I would like to be able to work in parallel while waiting for your implementation.
(A blog post/tutorial on guarded cubical type theory would also be welcome.)

If you have any suggestions on what I should do, they would be welcome.

**Comment by Saizan** [comment]
@xekoukou I've been wanting to merge for a while, there isn't anything major blocking but I need to find the time to do the necessary cleanup.

You can indeed prove bisimilar streams equal as shown in the cubical-demo repo: it works by the interaction of the interval and copatterns. It should also interact fine with Sized Types in case you need more flexible productivity checking.

Regarding "Guarded Cubical Type Theory" there are still some kinks to work out for that type theory so the implementation will have to wait a bit more.

**Comment by xekoukou** [comment]
@Saizan Nice, as soon as I finish some other work, I will look at your demos to get familiar with the techniques used.

I do not know the difference between Sized types and guarded recursion. But I am currently learning Sized types because it seems that I will need then.

**Comment by canndrew** [comment]
It looks like some of the cubical types stuff has made it on to master. Should this issue be closed now?

**Comment by Saizan** [comment]
Indeed, merged into master with commit

9bb7c9e07f500899b80f5dbcfe7717118be55d31
